src/engine/EntityReference.cs

```diff
@@ -0,0 +1,147 @@
+    private T currentInstance;
+    private AliveMarker currentAliveMarker;
+
+    public EntityReference(T value)
+    {
+        Value = value;
+    }
+
+    /// <summary>
+    ///   Creates a reference that doesn't refer to anything
+    /// </summary>
+    public EntityReference()
+    {
+    }
+
+    /// <summary>
+    ///   Gets the referenced entity if it is still alive
+    /// </summary>
+    /// <remarks>
+    ///   <para>
+    ///     If you need to access one object a bunch of times it's better to get the value from here once per frame
+    ///     and keep a normal reference around.
+    ///   </para>
+    /// </remarks>
+    public T Value
+    {
+        get
+        {
+            if (currentAliveMarker == null)
+                return null;
+
+            if (currentAliveMarker.Alive)
+                return currentInstance;
+
+            // Clear these references to let objects get garbage collected
+            currentInstance = null;
+            currentAliveMarker = null;
+            return null;
+        }
+        set
+        {
+            if (currentInstance == value)
+                return;
+
+            if (value == null)
+            {
+                currentInstance = null;
+                currentAliveMarker = null;
+            }
+            else
+            {
+                var marker = value.AliveMarker;
+
+                if (marker.Alive)
+                {
+                    currentInstance = value;
+                    currentAliveMarker = marker;
+                }
+                else
+                {
+                    currentInstance = null;
+                    currentAliveMarker = null;
+                }
+            }
+        }
+    }
+
+    [JsonIgnore]
+    public bool IsAlive => currentAliveMarker is { Alive: true };
+
+    public static implicit operator T(EntityReference<T> value)
+    {
+        return value.Value;
+    }
+
+    public static bool operator ==(EntityReference<T> lhs, EntityReference<T> rhs)
+    {
+        return Equals(lhs, rhs);
+    }
+
+    public static bool operator !=(EntityReference<T> lhs, EntityReference<T> rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    public static bool operator ==(EntityReference<T> lhs, T rhs)
+    {
+        if (lhs == null)
+            return rhs == null;
+
+        return lhs.Value == rhs;
+    }
+
+    public static bool operator !=(EntityReference<T> lhs, T rhs)
+    {
+        return !(lhs == rhs);
+    }
+
+    public void ClearIfNotAlive()
+    {
+        if (IsAlive)
+            return;
+
+        Value = null;
+    }
+
+    public override bool Equals(object obj)
+    {
+        if (obj == null)
+        {
+            return false;
+        }
+
+        if (GetType() != obj.GetType())
+        {
+            return false;
+        }
+
+        return Equals((EntityReference<T>)obj);
+    }
+
+    public bool Equals(EntityReference<T> obj)
+    {
+        return ReferenceEquals(Value, obj.Value);
```

Comment:
Would we consider references holding the same object, but with different alive values, equal? I would think we wouldn't.

Reply:
That would be a bug in the entity class. Entities are supposed to return the same alive instance each time the `AliveMarker` property is read. For speed reasons I'd prefer to not add a check about that here.